### ISSUE TITLE
Updating to Java 11

### ISSUE THREAD

**Issue by jhg023** [luna-rs/luna#82]
With support for Java 8 ending in January, 2019, I think it is a good time to discuss updating Luna to use Java 11. As Luna does not seem to be a commercial product, there shouldn't be an issue using Oracle's official JDK releases, even with their new licensing.

Besides the support deadline, there are **many** other reasons to update:
 

- Modularization (Java 9)
- G1 Garbage Collector (Java 9)
- Collection Factory Methods (Java 9)
- Type Inference Enhancements (Java 9)
- Latin1 Encoded `String`s (Java 9)
- GraalVM (Java 10)
- Local Variable Type Inference (Java 10)
- Built-in `HttpClient` (Java 11)
- An extensive amount of new methods (Java 9-11)

**Comment by tlf30** [comment]
I would recommend using the OpenJDK Java 11. Just to prevent issues with the license in the scenario that someone would like to use Luna for commercial use, or in the scenario that a commercial entity would like to run a community OSRS server. Not that I would recommend any commercial entity do this.

**Comment by jhg023** [comment]
I just realized that it probably doesn't matter which JDK we use to build it, so long that we don't create a self-contained executable with it. It would essentially be up to the user to select their JDK.

Side-note: I was able to build Luna on JDK 11, but I had to remove the following dependencies:

```xml
<dependency>
    <groupId>org.functionaljava</groupId>
    <artifactId>functionaljava-java8</artifactId>
    <version>4.7</version>
</dependency>
<dependency>
    <groupId>org.functionaljava</groupId>
    <artifactId>functionaljava-quickcheck</artifactId>
    <version>4.7</version>
</dependency>
<dependency>
    <groupId>org.functionaljava</groupId>
    <artifactId>functionaljava-java-core</artifactId>
    <version>4.7</version>
</dependency>```
No code changes were required (other than the addition of the `module-info.java`).

**Comment by lare96** [comment]
The functional java dependencies are only used in one or two situations (one of which Java 11 can replace) so I think we'd be okay removing them altogether. I think there are a few things that need to be considered when updating

- [ ] - Changing local variables to use 'var' syntax
- [ ] - Using Java 9 modules
- [ ] - Using new additions to the Optional API

Once issue #80 is completed I'll create a new branch for upgrading to Java 11. Do you guys have any suggestions or objections for code that should be updated while transitioning?

**Comment by jhg023** [comment]
@lare96 From what I've seen, the functional dependencies are only used for `<Integer, Integer>` tuples (which should definitely be their own class to prevent boxing) and to get a list of lines from a `String` (which is built in to Java 11 with `String#lines`).

I don't think any code modifications should occur until the server is successfully built on the new branch, but I do agree that we can take advantage of `var` (so long as we stick to the [style guide](https://openjdk.java.net/projects/amber/LVTIstyle.html)).

At first, the server should also be one big module and we can modularize it from there, if needed.

**Comment by jhg023** [comment]
It seems that the Scala dependencies that we rely on do not fully support Java 11 yet. Their next version milestones are set for late January and early February, 2019, so it would be best to put this issue on hold until then.

**Comment by lare96** [comment]
Ammonite unblocks this issue, but the only problem is we'll have to remain on Scala 2.12.6 until a version is compatible with Java 11

**Comment by AshotN** [comment]
JDK 11 doesn't seem to come with JavaFX anymore, or at least some parts of it. I followed this [Stackoverflow answer](https://stackoverflow.com/a/52470141/1009167) and download this [SDK](https://gluonhq.com/products/javafx/) to resolve the issues.

**Comment by lare96** [comment]
The included dependencies don't work? @AshotN  https://github.com/luna-rs/luna/blob/master/pom.xml#L143

**Comment by jhg023** [comment]
This issue was closed by #134
